sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -338,14 +369,20 @@ class SpillableHostBufferHandle private (
                 }
               }
             }
-            disk = Some(diskHandleBuilder.build)
-            sizeInBytes
+            val staging = Some(diskHandleBuilder.build)
+            synchronized {
+              if (!closed) {
+                disk = staging
+              }
+              releaseHostResource()
+            }
           }
-        } else {
-          0L
         }
+        sizeInBytes
+      } else {
+          0
       }
-      releaseHostResource()
+      toSpill = None
```

Comment:
I believe that this needs to be under a synchronized block to avoid race conditions.

Also because this can happen on a thread that didn't set  `toSpill` we might accidentally end up spilling twice. It should probably be moved up next to the withResource for `toSpill.get`, so that we set it to none after it is closed.